### PR TITLE
frr-isis: T1316: Add new routing feature ISIS

### PR DESCRIPTION
**Pull request by sever-sever**
Add ISIS routing feature

```
set protocols isis FOO interface eth1
set protocols isis FOO net '49.0000.0000.0000.0001.00'
set protocols isis FOO redistribute ipv4 connected level-2

```
```
vyos@r4-roll# run show ip route isis
Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric,
       > - selected route, * - FIB route, q - queued route, r - rejected route

I   100.64.0.0/24 [115/20] via 100.64.0.1, eth1 inactive, 00:06:07
I>* 100.64.222.0/24 [115/65] via 100.64.0.1, eth1, 00:06:07
I   192.168.122.0/24 [115/10] via 100.64.0.1, eth1, 00:06:07
I>* 203.0.113.0/24 [115/65] via 100.64.0.1, eth1, 00:06:07

```

### CODE REVIEW

**Review by dmbaturin — CHANGES_REQUESTED** [review]
Can we rename the node to "is-is"?
There are two reasons. First, it's the official spelling of the [RFC](https://tools.ietf.org/html/rfc1142) and the ISO standard.

Second, "ISIS" spelling is now so firmly associated with terrorism that we should avoid it if we can. ;)

**Comment by sever-sever** [comment]
> Can we rename the node to "is-is"?

In cisco and juniper this feature also named "isis"

cisco
```
cisco(config)#router ?
  bgp       Border Gateway Protocol (BGP)
  egp       Exterior Gateway Protocol (EGP)
  eigrp     Enhanced Interior Gateway Routing Protocol (EIGRP)
  isis      ISO IS-IS
```
juniper
```
mx80# set protocols i
                                    ^
'i' is ambiguous.
Possible completions:
> iccp                 ICCP options
> igmp                 IGMP options
> igmp-snooping        IGMP snooping configuration
> ilmi                 Interim Local Management Interface Protocol configuration
> isis                 IS-IS options

```
Regarding RFC
The word "is-is" occurs 3 times.
The word "isis" occurs 512 times.

We have to think.

**Comment by sever-sever** [comment]
@dmbaturin done, is-is.

File: interface-definitions/protocols-isis.xml.in [hunk 1]
```diff
@@ -3,9 +3,9 @@
 <interfaceDefinition>
   <node name="protocols">
     <children>
-      <tagNode name="isis" owner="${vyos_conf_scripts_dir}/protocols_isis.py">
```
Comment (c-po) [inline interface-definitions/protocols-isis.xml.in:6]: Why is this renamed to is-is? All Vendors I know call it isis, also FRR. We should stick to the industry standard.

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
Well done @sever-sever

**Comment by c-po** [comment]
All vendors I know call it isis, also FRR. We should stick to the industry standard.

Thanks for counting the number occurances on it.

I vote for isis, if we can cast a vote.

**Comment by Cheeze-It** [comment]
Juniper calls it "isis":
https://www.juniper.net/documentation/en_US/junos/topics/example/routing-protocol-is-is-security-configuring-cli.html

Cisco calls it "isis":
https://www.cisco.com/c/en/us/support/docs/ip/integrated-intermediate-system-to-intermediate-system-is-is/13795-is-is-ip-config.html

Nokia/Alcatel Lucent calls it "isis":
https://infocenter.nokia.com/public/770562R1A/index.jsp?topic=%2Fcom.sar.routing_protocols%2Fhtml%2Fisis_config.html

HP Comware calls it "isis":
https://support.hpe.com/hpesc/public/docDisplay?docId=emr_na-c03189554

Now, my personal desire is to use "isis." I personally am ok with either but if I could pick/choose I'd choose "isis."

**Comment by dmbaturin** [comment]
The question is _why_ they call it "ISIS" without a hyphen... it may be purely historical reasons. 

## ISO standard spelling

IS-IS is an ISO standard that was given an RFC status by IETF, rather than developed via the normal community process. 

The current [ISO standard](https://standards.iso.org/ittf/PubliclyAvailableStandards/c030932_ISO_IEC_10589_2002(E).zip) (second edition, 2002) uses "IS-IS" _consistently_ when it refers to the protocol name. E.g. "The Intradomain IS-IS protocol exploits multicast capabilities for all IS-IS protocol exchanges on broadcast subnetworks."
It only uses "ISIS" spelling in field names like `iSISHelloTimer`, presumably to allow fields to be named consistently in all programming languages, since most of them don't allow hyphens in variable names.

Regarding counting occurences... Note that the ASCII version of that RFC has the following warning:

```
NOTE:  This is a bad ASCII version of this document.  The official
document is the PostScript file, which has the diagrams in place.
Please use the PostScript version of this memo.
```

This is very plausible: its typesetting quality is indeed much lower than the normal "born-ASCII" RFCs that include ASCII diagrams among other things.

Here's the PDF version of that RFC: https://www.rfc-editor.org/rfc/rfc1142.pdf
It appears to be an "RFC-ified" copy of the original ISO standard, and it also uses "IS-IS" throughout (other than in field names).

I think this proves that the official spelling was always supposed to be "IS-IS", and "ISIS" in the document body is a transcription error (quite likely—even now automatic extraction of text from PDFs isn't always flawless).

## Spelling in other vendors' documentation

Another point is that JunOS uses `isis` in the CLI, but their documentation uses "IS-IS" a lot: https://www.juniper.net/documentation/en_US/junos/topics/concept/is-is-routing-overview.html
And so does Cisco: https://www.cisco.com/c/en/us/support/docs/ip/integrated-intermediate-system-to-intermediate-system-is-is/13795-is-is-ip-config.html

Thus, if we use `is-is` in the CLI, we can keep the CLI and the docs consistent with the ISO standard—unlike other vendors. ;)

I don't know why they created that inconsistency and why they are tolerating it, but it seems like if we copy it, we'll be copying a mistake.

## Conclusion

* "IS-IS" spelling is used consistently in the ISO standard and the original PDF version of the RFC.
* Other vendors use "IS-IS" in the documentation even though it's `isis` in their CLI—an inconsistency we can avoid at no cost.
* For users, it doesn't matter, since no other protocol starts with "is" and it's going to complete to the right word either way.
* Some people in Phabricator did request that we use "IS-IS" specifically because "ISIS" is now associated with the terrorist organization, and they have a point.

**Comment by jack9603301** [comment]
When an implementation conflicts with a standard, do I think the standard naming should be followed first?

**Comment by Cheeze-It** [comment]
A lot of this unfortunately comes down to historical reasons on the why. I speculate that the name within the CLI was taken what it is because the extra keystrokes initially to put down "is-is" are going to be more cumbersome than "isis." The other thing that I kinda would worry about would be automatability. I am unsure if the hyphen would cause problems with certain automation frameworks. I am unsure if Jinja2 would have a problem with a hypen. I'm thinking not but it's always something to keep in mind. 

When it comes to standards, a lot of times there's a divergence between documentation of a protocol vs the configuration trees of a device. The standards themselves are not as set in stone as a lot of people would like to believe unfortunately. 

>> I don't know why they created that inconsistency and why they are tolerating it, but it seems like if we copy it, we'll be copying a mistake.

This mostly has to do with ease of use and nomenclature. In the network space, divergence is generally frowned upon by engineers or administrators. Mostly because CLI parsers are hard to memorize. It's one of the reasons why all the vendors converged around a spelling.

**Comment by dmbaturin** [comment]
Since the ISIS spelling won the polls, shall we use that spelling? @sever-sever could you revert it back to `isis`?

**Comment by sever-sever** [comment]
> Since the ISIS spelling won the polls, shall we use that spelling? @sever-sever could you revert it back to `isis`?

@dmbaturin  Done

**Pull request merged**